File: apps/api/src/routes/public.ts
import { and, desc, eq, lt } from 'drizzle-orm';
import { Hono } from 'hono';
import type { GalleryResponse } from '@leetete/shared';
import { getDb } from '../db/client.js';
import { eventConfig, uploads } from '../db/schema.js';
import type { Bindings } from '../env.js';
import { generateQrPdf } from '../lib/qr-pdf.js';
import { generateQrPng, generateQrSvg } from '../lib/qrcode.js';
import { getStorage } from '../lib/storage-factory.js';

export const publicRoutes = new Hono<Bindings>();

const GALLERY_PAGE_SIZE = 24;

publicRoutes.get('/event', async (c) => {
  const db = getDb();
  const cfg = await db.select().from(eventConfig).where(eq(eventConfig.id, 1));
  const row = cfg[0];
  if (!row) return c.json({ error: 'not_configured' }, 404);
  const storage = getStorage();
  return c.json({
    coupleNames: row.coupleNames,
    eventDate: row.eventDate,
    welcomeMessage: row.welcomeMessage,
    galleryVisibility: row.galleryVisibility,
    allowVideo: row.allowVideo,
    maxFileMb: row.maxFileMb,
    maxVideoSeconds: row.maxVideoSeconds,
    coverUrl: row.coverKey ? storage.publicUrl(row.coverKey) : null,
  });
});

publicRoutes.get('/gallery', async (c) => {
  const db = getDb();
  const cfg = await db.select().from(eventConfig).where(eq(eventConfig.id, 1));
  const row = cfg[0];
  if (!row || row.galleryVisibility !== 'public') {
    return c.json({ items: [], nextCursor: null } satisfies GalleryResponse);
  }

  const cursorParam = c.req.query('cursor');
  const cursor = cursorParam ? Number(cursorParam) : null;
  const limit = Math.min(Number(c.req.query('limit') ?? GALLERY_PAGE_SIZE), 60);

  const baseCondition = eq(uploads.status, 'approved');
  const where =
    cursor !== null && Number.isFinite(cursor)
      ? and(baseCondition, lt(uploads.createdAt, cursor))
      : baseCondition;

  const rows = await db
    .select()
    .from(uploads)
    .where(where)
    .orderBy(desc(uploads.createdAt))
    .limit(limit + 1);

  const hasMore = rows.length > limit;
  const slice = hasMore ? rows.slice(0, limit) : rows;
  const storage = getStorage();

  const items = slice.map((r) => ({
    id: r.id,
    url: storage.publicUrl(r.storageKey),
    thumbnailUrl: r.thumbnailKey ? storage.publicUrl(r.thumbnailKey) : null,
    mimeType: r.mimeType,
    isVideo: r.mimeType.startsWith('video/'),
    authorName: r.authorName,
    message: r.message,
    createdAt: r.createdAt,
  }));

  const nextCursor = hasMore ? String(slice[slice.length - 1]!.createdAt) : null;
  return c.json({ items, nextCursor } satisfies GalleryResponse);
});

publicRoutes.get('/qrcode', async (c) => {
  const format = (c.req.query('format') ?? 'png').toLowerCase();
  const overrideUrl = c.req.query('url');

  const reqUrl = new URL(c.req.url);
  const base = overrideUrl ?? c.env.PUBLIC_BASE_URL ?? `${reqUrl.protocol}//${reqUrl.host}`;
  const target = base.replace(/\/$/, '') + '/enviar';

  if (format === 'svg') {
    const svg = await generateQrSvg(target);
    return new Response(svg, {
      headers: {
        'content-type': 'image/svg+xml; charset=utf-8',
        'cache-control': 'public, max-age=300',
      },
    });
  }

  if (format === 'pdf') {
    const db = getDb();
    const cfg = await db.select().from(eventConfig).where(eq(eventConfig.id, 1));
    const row = cfg[0];
    const pdf = await generateQrPdf({
      url: target,
      coupleNames: row?.coupleNames ?? c.env.COUPLE_NAMES,
      eventDate: row?.eventDate ?? c.env.EVENT_DATE,
    });
    return new Response(pdf, {
      headers: {
        'content-type': 'application/pdf',
        'content-disposition': 'inline; filename="qr-mesa.pdf"',
        'cache-control': 'no-store',
      },
    });
  }

  const png = await generateQrPng(target, 800);
  return new Response(png, {
    headers: {
      'content-type': 'image/png',
      'cache-control': 'public, max-age=300',
    },
  });
});

publicRoutes.get('/stats', async (c) => {
  const db = getDb();
  const all = await db
    .select({ id: uploads.id, mimeType: uploads.mimeType })
    .from(uploads)
    .where(eq(uploads.status, 'approved'));
  const photos = all.filter((u) => u.mimeType.startsWith('image/')).length;
  const videos = all.filter((u) => u.mimeType.startsWith('video/')).length;
  return c.json({ photos, videos, total: all.length });
});

File: apps/api/app/routes/public.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.responses import Response
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..config import settings
from ..db.base import get_db
from ..db.models import EventConfig, Upload
from ..lib.pdf import generate_qr_pdf
from ..lib.qrcode_gen import generate_qr_png, generate_qr_svg
from ..lib.storage import storage

router = APIRouter()
GALLERY_PAGE_SIZE = 24

Db = Annotated[AsyncSession, Depends(get_db)]


@router.get("/event")
async def get_event(db: Db):
    cfg = (
        await db.execute(select(EventConfig).where(EventConfig.id == 1))
    ).scalar_one_or_none()
    if not cfg:
        raise HTTPException(404, detail="not_configured")
    return {
        "coupleNames": cfg.couple_names,
        "eventDate": cfg.event_date,
        "welcomeMessage": cfg.welcome_message,
        "galleryVisibility": cfg.gallery_visibility,
        "allowVideo": cfg.allow_video,
        "maxFileMb": cfg.max_file_mb,
        "maxVideoSeconds": cfg.max_video_seconds,
        "coverUrl": storage.public_url(cfg.cover_key) if cfg.cover_key else None,
    }


@router.get("/gallery")
async def get_gallery(
    db: Db,
    cursor: str | None = None,
    limit: Annotated[int, Query(le=60, ge=1)] = GALLERY_PAGE_SIZE,
):
    cfg = (
        await db.execute(select(EventConfig).where(EventConfig.id == 1))
    ).scalar_one_or_none()
    if not cfg or cfg.gallery_visibility != "public":
        return {"items": [], "nextCursor": None}

    cursor_val: int | None = None
    if cursor and cursor.isdigit():
        cursor_val = int(cursor)

    stmt = select(Upload).where(Upload.status == "approved")
    if cursor_val is not None:
        stmt = stmt.where(Upload.created_at < cursor_val)
    stmt = stmt.order_by(Upload.created_at.desc()).limit(limit + 1)

    rows = list((await db.execute(stmt)).scalars().all())
    has_more = len(rows) > limit
    slice_rows = rows[:limit] if has_more else rows

    items = [
        {
            "id": r.id,
            "url": storage.public_url(r.storage_key),
            "thumbnailUrl": storage.public_url(r.thumbnail_key)
            if r.thumbnail_key
            else None,
            "mimeType": r.mime_type,
            "isVideo": r.mime_type.startswith("video/"),
            "authorName": r.author_name,
            "message": r.message,
            "createdAt": r.created_at,
        }
        for r in slice_rows
    ]
    next_cursor = str(slice_rows[-1].created_at) if has_more and slice_rows else None
    return {"items": items, "nextCursor": next_cursor}


@router.get("/qrcode")
async def get_qrcode(
    request: Request,
    db: Db,
    format: str = "png",
    url: str | None = None,
):
    fmt = format.lower()
    if url:
        base = url
    elif settings.PUBLIC_BASE_URL:
        base = settings.PUBLIC_BASE_URL
    else:
        base = str(request.base_url).rstrip("/")
    target = base.rstrip("/") + "/enviar"

    if fmt == "svg":
        svg = generate_qr_svg(target)
        return Response(
            content=svg,
            media_type="image/svg+xml; charset=utf-8",
            headers={"cache-control": "public, max-age=300"},
        )

    if fmt == "pdf":
        cfg = (
            await db.execute(select(EventConfig).where(EventConfig.id == 1))
        ).scalar_one_or_none()
        pdf_bytes = generate_qr_pdf(
            target,
            cfg.couple_names if cfg else settings.COUPLE_NAMES,
            cfg.event_date if cfg else settings.EVENT_DATE,
        )
        return Response(
            content=pdf_bytes,
            media_type="application/pdf",
            headers={
                "content-disposition": 'inline; filename="qr-mesa.pdf"',
                "cache-control": "no-store",
            },
        )

    png = generate_qr_png(target, 800)
    return Response(
        content=png,
        media_type="image/png",
        headers={"cache-control": "public, max-age=300"},
    )


@router.get("/stats")
async def get_stats(db: Db):
    rows = (
        await db.execute(
            select(Upload.id, Upload.mime_type).where(Upload.status == "approved")
        )
    ).all()
    photos = sum(1 for r in rows if r.mime_type.startswith("image/"))
    videos = sum(1 for r in rows if r.mime_type.startswith("video/"))
    return {"photos": photos, "videos": videos, "total": len(rows)}
